Prevented HSQL Exception from Being Thrown on Startup when Checking if the Server Is Already Running

nathandmoore opened the pull request:
Original Slack conversation: https://blc.slack.com/archives/C3ZH6QRAQ/p1490987962636341

On startup and shutdown, the following exception is logged by HSQL:

```java
org.hsqldb.HsqlException: Network client is not a HSQLDB JDBC driver
    at org.hsqldb.error.Error.error(Unknown Source)
    at org.hsqldb.error.Error.error(Unknown Source)
    at org.hsqldb.server.ServerConnection.handshake(Unknown Source)
    at org.hsqldb.server.ServerConnection.init(Unknown Source)
    at org.hsqldb.server.ServerConnection.run(Unknown Source)
    at java.lang.Thread.run(Thread.java:745)
```

This is also logged when you startup the "other" app (i.e., if you startup admin first, then this error is logged when you start site, but it will appear it admin's logs not site's). It is caused by the way we were checking if the HSQL server was already running:

```java
try (Socket ignored = new Socket(InetAddress.getByName(null), props.getIntegerProperty("server.port", 0)))
```

This socket is, in fact, *not* an "HSQLDB JDBC driver". Therefore, to prevent this exception we should use a `java.sql.Connection` to test if the server is running instead. That way, the HSQL JDBC driver *will* be used when creating the connection. Also, we need to log an error if the connection is refused because the port we are trying to setup the HSQL server on is already in use.

I also added logging indicating whether the HSQL JDBC driver is missing and, if the server is already running, that we are not creating a duplicate server.

## Testing

1. Startup admin

Should not see the afore mentioned exception, instead you should see:

```java
2017-08-02 11:09:36.768  INFO 51223 --- [ost-startStop-1] c.b.autoconfigure.HSQLDBServer           : HSQL server is not running.
2017-08-02 11:09:36.769  INFO 51223 --- [ost-startStop-1] c.b.autoconfigure.HSQLDBServer           : Starting HSQL server...
2017-08-02 11:09:36.769  WARN 51223 --- [ost-startStop-1] c.b.autoconfigure.HSQLDBServer           : HSQL embedded database server is for demonstration purposes only and is not intended for production usage.
```

2. Start site:

```java
2017-08-02 11:12:05.689  INFO 51561 --- [ost-startStop-1] c.b.autoconfigure.HSQLDBServer           : HSQL server is already running. Will not start a new instance.
```

3. Shutdown site
    - Should see not exception

nathandmoore committed:
Trying to open a socket causes any previously openned connections to log an exception because the socket does not use the HSQL JDBC driver. Instead, we should use the standard way to test a DB connection by creating a `java.sql.Connection`, which uses the JDBC driver for HSQL to open a connection. We can then close the connection if successful.

nathandmoore committed:
If the connection is refused, make sure that a server isn't already using that port. If so, then log an error.

nathandmoore commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(0.5) Look into suppressing/fixing the org.hsqldb.HsqlException: Network client is not a HSQLDB JDBC driver error message on startup](https://trello.com/c/fssXR9fn/64-05-look-into-suppressing-fixing-the-orghsqldbhsqlexception-network-client-is-not-a-hsqldb-jdbc-driver-error-message-on-startup)

phillipuniverse commented on src/main/java/com/broadleafcommerce/autoconfigure/HSQLDBServer.java:
```diff
@@ -64,20 +68,62 @@ public HSQLDBServer(final HSQLDBProperties autoProps, Environment environment) {
+
+        final int port = props.getIntegerProperty("server.port", 0);
+        try {
+            Class.forName("org.hsqldb.jdbc.JDBCDriver");
```
Since we are actually within the HSQLDB starter, this check is not necessary.

phillipuniverse commented on src/main/java/com/broadleafcommerce/autoconfigure/HSQLDBServer.java:
```diff
@@ -64,20 +68,62 @@ public HSQLDBServer(final HSQLDBProperties autoProps, Environment environment) {
+            isRunning = true;
+        } catch (ClassNotFoundException e) {
+            LOG.error("ERROR: Cannot test if an HSQL server is running: failed to load HSQLDB JDBC driver.", e);
+        } catch (SQLException e) {
```
Do we get the same exception if something is there and not responding correctly vs something isn't there at all?

nathandmoore replied:
Yes

phillipuniverse commented on src/main/java/com/broadleafcommerce/autoconfigure/HSQLDBServer.java:
```diff
@@ -64,20 +68,62 @@ public HSQLDBServer(final HSQLDBProperties autoProps, Environment environment) {
+            try {
+                // see if the port is being used already (by something other than HSQL)
+                s = new Socket(InetAddress.getByName(null), port);
+                LOG.error("Port," + port + ", is already in use but not by HSQL");
```
In this log message can you add more information about what the user can do to resolve the problem? Essentially how to check what is consuming that port.

nathandmoore replied:
So, tell them to use something like `netstat -vanl | grep $PORT_ID`?  Although, I don't know what that does on windows. Might have to give a few examples for different platforms.

phillipuniverse replied:
Right. Just trying to solve the _next_ problem they will have. The next thing anybody would do at this point would Google "How do I find out what process is running on a port?" and if we put that front and center it gives them a better experience overall. Basically just think what _you_ would want to see if you got this error. What would be most helpful to you? (and think about windows/linux/macos).

phillipuniverse commented on src/main/java/com/broadleafcommerce/autoconfigure/HSQLDBServer.java:
```diff
@@ -64,20 +68,62 @@ public HSQLDBServer(final HSQLDBProperties autoProps, Environment environment) {
+                closeConnectionQuietly(s);
+            }
+        } catch (UnknownHostException e) {
+            LOG.error("ERROR: An error occurred while testing whether an HSQL server is running.", e);
```
What's the use case of actually getting an `UnknownHostException`? Should we actually log an error here? Is there any way for the user to recover?

nathandmoore replied:
If we assume we are using localhost (and not using `InetAddress.getByName()`) then we don't need to have this.

phillipuniverse commented on src/main/java/com/broadleafcommerce/autoconfigure/HSQLDBServer.java:
```diff
@@ -64,20 +68,62 @@ public HSQLDBServer(final HSQLDBProperties autoProps, Environment environment) {
+        final int port = props.getIntegerProperty("server.port", 0);
+        try {
+            Class.forName("org.hsqldb.jdbc.JDBCDriver");
+            final String url = "jdbc:hsqldb:hsql://" + InetAddress.getByName(null).getHostName() + ":"
```
I am not sure how reliable `InetAddress.getByName()` actually is. There are instances where you might not actually have this host in `/etc/hosts` and this particular call causes the entire startup to hang for basically forever (or like 5 minutes). I have personally seen this happen on OSX when I did not have the IPV6 entry for `localhost` in my `/etc/hosts`.

On VMs on various cloud providers I have also seen entries in `/etc/hosts` that map the host name to the _external_ IP address which means all traffic has to go out and back in again... not great.

Since this whole project assumes that HSQL and everything else is local, we should just use `127.0.0.1` since there is no OS-level translation that has to go on with host names.

phillipuniverse commented on src/main/java/com/broadleafcommerce/autoconfigure/HSQLDBServer.java:
```diff
@@ -64,20 +68,62 @@ public HSQLDBServer(final HSQLDBProperties autoProps, Environment environment) {
+        return isRunning;
+    }
+    
+    protected void closeConnectionQuietly(AutoCloseable c) {
```
Previously, I was using the Java 7-compatible [try with resource](https://www.mkyong.com/java/try-with-resources-example-in-jdk-7/) that does exactly what you're doing here with anything that implements `AutoCloseable`. Do that instead of this extra call in a `finally` block.

nathandmoore committed:
- Reduced the number of exceptions handled since several should not occur in this context.
    -Since we are in the HSQLDB starter, we can assume that the correct JDBC driver will be available.
    - Moreover, we can assume that we will be using localhost for embedded HSQLDB.
- Used "try with resource" instead of manually closing the connection or socket
- Added more detail to the log message for how to find out which process is using the HSQL port (when not HSQL).

nathandmoore committed:
Switched to the actual address of localhost in case localhost isn't mapped correctly (some cloud instances) and to remove the need to resolve the hostname to it's actual address.
Per https://github.com/BroadleafCommerce/HSQLDatabaseStarter/pull/1#discussion_r130983937

phillipuniverse commented:
LGTM great work!
